File: backend/src/services/reviewService.js
const Review = require("../models/Review");
const Profile = require("../models/Profile");
const Appointment = require("../models/Appointment");

async function createReview(patientId, therapistId, rating, comment) {
	if (!therapistId || rating == null) {
		const err = new Error("therapistId and rating are required");
		err.status = 400;
		throw err;
	}

	const hadConfirmed = await Appointment.exists({
		pt: therapistId,
		patient: patientId,
		status: "confirmed",
	});
	if (!hadConfirmed) {
		const err = new Error(
			"Only patients with a confirmed appointment can leave a review",
		);
		err.status = 403;
		throw err;
	}

	const review = new Review({
		therapist: therapistId,
		patient: patientId,
		rating,
		comment,
	});
	await review.save();

	const agg = await Review.aggregate([
		{ $match: { therapist: review.therapist } },
		{ $group: { _id: "$therapist", avgRating: { $avg: "$rating" } } },
	]);
	const avgRating = agg[0] ? agg[0].avgRating : rating;

	await Profile.findOneAndUpdate(
		{ user: therapistId },
		{ rating: avgRating },
		{ upsert: false },
	);

	// Return populated review for API consumers
	const populated = await Review.findById(review._id)
		.populate({
			path: "therapist",
			select: "profile",
			populate: { path: "profile" },
		})
		.populate({
			path: "patient",
			select: "profile",
			populate: { path: "profile" },
		});
	return populated;
}

async function getReviewsForTherapist(therapistId) {
	const reviews = await Review.find({ therapist: therapistId }).populate(
		"patient",
		"profile",
	);
	return reviews;
}

module.exports = {
	createReview,
	getReviewsForTherapist,
};

File: backend/internal/service/review_service.go
package service

import (
	"context"
	"database/sql"
	"strings"

	"github.com/google/uuid"

	"github.com/divijg19/physiolink/backend/internal/db"
)

type ReviewService struct{ db *db.DB }

func NewReviewService(d *db.DB) *ReviewService { return &ReviewService{db: d} }

// CreateReview enforces that the patient has at least one appointment with the therapist
// before allowing a review. It links the review to the most recent appointment between them.
func (s *ReviewService) CreateReview(ctx context.Context, patientID, therapistID uuid.UUID, rating int, comment string) (map[string]interface{}, error) {
	// find recent appointment
	var apptID uuid.UUID
	err := s.db.Pool.QueryRow(ctx, `SELECT id FROM appointments
        WHERE patient_id=$1 AND therapist_id=$2
        ORDER BY created_at DESC LIMIT 1`, patientID, therapistID).Scan(&apptID)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, &ForbiddenError{Msg: "Only patients with an appointment can leave a review"}
		}
		return nil, err
	}

	var reviewID uuid.UUID
	err = s.db.Pool.QueryRow(ctx, `INSERT INTO reviews (appointment_id, patient_id, rating, comment)
        VALUES ($1,$2,$3,$4) RETURNING id`, apptID, patientID, rating, comment).Scan(&reviewID)
	if err != nil {
		return nil, err
	}

	// Update average rating on therapist profile
	var avg sql.NullFloat64
	_ = s.db.Pool.QueryRow(ctx, `SELECT AVG(r.rating)::float
		FROM appointments a JOIN reviews r ON r.appointment_id = a.id
		WHERE a.therapist_id = $1`, therapistID).Scan(&avg)
	if avg.Valid {
		_, _ = s.db.Pool.Exec(ctx, `UPDATE profiles SET rating = $2 WHERE user_id = $1`, therapistID, avg.Float64)
	}

	out := map[string]interface{}{
		"_id":       reviewID.String(),
		"therapist": map[string]interface{}{"_id": therapistID.String()},
		"patient":   map[string]interface{}{"_id": patientID.String()},
		"rating":    rating,
		"comment":   comment,
	}
	return out, nil
}

func (s *ReviewService) GetReviewsForTherapist(ctx context.Context, therapistID uuid.UUID) ([]map[string]interface{}, error) {
	rows, err := s.db.Queries.GetReviewsForTherapist(ctx, therapistID)
	if err != nil {
		return nil, err
	}
	var out []map[string]interface{}
	for _, r := range rows {
		firstName, lastName := "", ""
		if r.PatientName.Valid {
			displayName := r.PatientName.String
			if idx := strings.Index(displayName, " "); idx > 0 {
				firstName = displayName[:idx]
				lastName = strings.TrimSpace(displayName[idx+1:])
			} else {
				firstName = displayName
			}
		}
		patient := map[string]interface{}{
			"_id": r.PatientID.String(),
			"profile": map[string]interface{}{
				"firstName": firstName,
				"lastName":  lastName,
			},
		}
		m := map[string]interface{}{
			"_id":     r.ReviewID,
			"patient": patient,
			"rating":  r.Rating,
		}
		if r.Comment.Valid {
			m["comment"] = r.Comment.String
		}
		out = append(out, m)
	}
	return out, nil
}

type ForbiddenError struct{ Msg string }

func (e *ForbiddenError) Error() string { return e.Msg }
